File: sdk/typescript/src/git.ts
/**
 * Git link functionality for AgentTrace.
 *
 * Git links allow you to associate traces and observations with git commits,
 * branches, and repositories for better traceability and debugging.
 */

import { execSync } from "child_process";
import type { AgentTrace } from "./client";

export type GitLinkType = "start" | "commit" | "restore" | "branch" | "diff";

export interface GitLinkOptions {
  type?: GitLinkType;
  observationId?: string;
  commitSha?: string;
  branch?: string;
  repoUrl?: string;
  commitMessage?: string;
  filesChanged?: string[];
  autoDetect?: boolean;
}

export interface GitLinkInfo {
  id: string;
  traceId: string;
  observationId?: string;
  type: GitLinkType;
  commitSha: string;
  branch?: string;
  repoUrl?: string;
  commitMessage?: string;
  authorName?: string;
  authorEmail?: string;
  filesChanged: string[];
  createdAt: Date;
}

/**
 * Client for creating and managing git links.
 *
 * @example
 * ```typescript
 * const client = new AgentTrace({ apiKey: "..." });
 * const trace = client.trace({ name: "my-agent" });
 *
 * // Auto-link current git state
 * trace.gitLink();
 *
 * // Or with explicit info
 * trace.gitLink({
 *   commitSha: "abc123",
 *   branch: "main",
 *   type: "commit"
 * });
 * ```
 */
export class GitClient {
  private _client: AgentTrace;

  constructor(client: AgentTrace) {
    this._client = client;
  }

  /**
   * Create a git link for a trace.
   */
  link(traceId: string, options: GitLinkOptions = {}): GitLinkInfo {
    const linkId = generateId();
    const now = new Date();
    const type = options.type || "commit";
    const autoDetect = options.autoDetect !== false;

    let commitSha = options.commitSha;
    let branch = options.branch;
    let repoUrl = options.repoUrl;
    let commitMessage = options.commitMessage;
    let authorName: string | undefined;
    let authorEmail: string | undefined;
    let filesChanged = options.filesChanged;

    // Auto-detect git info
    if (autoDetect) {
      const gitInfo = this.getGitInfo();
      if (!commitSha) commitSha = gitInfo.commitSha;
      if (!branch) branch = gitInfo.branch;
      if (!repoUrl) repoUrl = gitInfo.repoUrl;
      if (!commitMessage) commitMessage = gitInfo.commitMessage;

      const authorInfo = this.getAuthorInfo();
      authorName = authorInfo.name;
      authorEmail = authorInfo.email;

      if (!filesChanged) {
        filesChanged = this.getChangedFiles();
      }
    }

    filesChanged = filesChanged || [];
    commitSha = commitSha || "";

    // Calculate diff stats
    const diffStats = this.getDiffStats();

    // Send to API
    if (this._client.enabled) {
      this._client._addEvent({
        type: "git-link-create",
        body: {
          id: linkId,
          traceId,
          observationId: options.observationId,
          linkType: type,
          commitSha,
          branch,
          repoUrl,
          commitMessage,
          authorName,
          authorEmail,
          filesChanged,
          additions: diffStats.additions,
          deletions: diffStats.deletions,
          timestamp: now.toISOString(),
        },
      });
    }

    return {
      id: linkId,
      traceId,
      observationId: options.observationId,
      type,
      commitSha,
      branch,
      repoUrl,
      commitMessage,
      authorName,
      authorEmail,
      filesChanged,
      createdAt: now,
    };
  }

  /**
   * Get current git repository info.
   */
  getGitInfo(): {
    commitSha?: string;
    branch?: string;
    repoUrl?: string;
    commitMessage?: string;
  } {
    const result: {
      commitSha?: string;
      branch?: string;
      repoUrl?: string;
      commitMessage?: string;
    } = {};

    try {
      result.commitSha = execSync("git rev-parse HEAD", { encoding: "utf-8", timeout: 5000 }).trim();
    } catch {
      // Not in a git repo
    }

    try {
      result.branch = execSync("git rev-parse --abbrev-ref HEAD", { encoding: "utf-8", timeout: 5000 }).trim();
    } catch {
      // Ignore
    }

    try {
      result.repoUrl = execSync("git config --get remote.origin.url", { encoding: "utf-8", timeout: 5000 }).trim();
    } catch {
      // Ignore
    }

    try {
      result.commitMessage = execSync('git log -1 --format=%s', { encoding: "utf-8", timeout: 5000 }).trim();
    } catch {
      // Ignore
    }

    return result;
  }

  /**
   * Get author info for current commit.
   */
  getAuthorInfo(): { name?: string; email?: string } {
    const result: { name?: string; email?: string } = {};

    try {
      result.name = execSync('git log -1 --format=%an', { encoding: "utf-8", timeout: 5000 }).trim();
    } catch {
      // Ignore
    }

    try {
      result.email = execSync('git log -1 --format=%ae', { encoding: "utf-8", timeout: 5000 }).trim();
    } catch {
      // Ignore
    }

    return result;
  }

  /**
   * Get list of changed files (staged and unstaged).
   */
  getChangedFiles(): string[] {
    const files: string[] = [];

    try {
      const staged = execSync("git diff --cached --name-only", { encoding: "utf-8", timeout: 5000 });
      files.push(...staged.trim().split("\n").filter(Boolean));
    } catch {
      // Ignore
    }

    try {
      const unstaged = execSync("git diff --name-only", { encoding: "utf-8", timeout: 5000 });
      for (const f of unstaged.trim().split("\n").filter(Boolean)) {
        if (!files.includes(f)) {
          files.push(f);
        }
      }
    } catch {
      // Ignore
    }

    try {
      const untracked = execSync("git ls-files --others --exclude-standard", { encoding: "utf-8", timeout: 5000 });
      for (const f of untracked.trim().split("\n").filter(Boolean)) {
        if (!files.includes(f)) {
          files.push(f);
        }
      }
    } catch {
      // Ignore
    }

    return files;
  }

  /**
   * Get diff statistics.
   */
  getDiffStats(): { additions: number; deletions: number } {
    const result = { additions: 0, deletions: 0 };

    try {
      const output = execSync("git diff --shortstat", { encoding: "utf-8", timeout: 5000 }).trim();
      // Parse: "2 files changed, 10 insertions(+), 5 deletions(-)"
      const insertMatch = output.match(/(\d+) insertion/);
      const deleteMatch = output.match(/(\d+) deletion/);
      if (insertMatch) result.additions = parseInt(insertMatch[1], 10);
      if (deleteMatch) result.deletions = parseInt(deleteMatch[1], 10);
    } catch {
      // Ignore
    }

    return result;
  }
}

function generateId(): string {
  if (typeof crypto !== "undefined" && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

File: sdk/go/git.go
package agenttrace

import (
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
)

// GitLinkType represents the type of git link.
type GitLinkType string

const (
	GitLinkTypeStart   GitLinkType = "start"
	GitLinkTypeCommit  GitLinkType = "commit"
	GitLinkTypeRestore GitLinkType = "restore"
	GitLinkTypeBranch  GitLinkType = "branch"
	GitLinkTypeDiff    GitLinkType = "diff"
)

// GitLinkOptions holds options for creating a git link.
type GitLinkOptions struct {
	Type          GitLinkType
	ObservationID string
	CommitSha     string
	Branch        string
	RepoURL       string
	CommitMessage string
	FilesChanged  []string
	AutoDetect    bool
}

// GitLinkInfo contains information about a created git link.
type GitLinkInfo struct {
	ID            string
	TraceID       string
	ObservationID string
	Type          GitLinkType
	CommitSha     string
	Branch        string
	RepoURL       string
	CommitMessage string
	AuthorName    string
	AuthorEmail   string
	FilesChanged  []string
	CreatedAt     time.Time
}

// GitLink creates a git link for this trace.
func (t *Trace) GitLink(opts *GitLinkOptions) *GitLinkInfo {
	if opts == nil {
		opts = &GitLinkOptions{}
	}

	linkID := uuid.New().String()
	now := time.Now().UTC()

	if opts.Type == "" {
		opts.Type = GitLinkTypeCommit
	}

	// Auto-detect by default
	autoDetect := opts.AutoDetect || (opts.CommitSha == "" && opts.Branch == "")

	var commitSha, branch, repoURL, commitMessage string
	var authorName, authorEmail string
	var filesChanged []string

	if autoDetect {
		gitInfo := getGitInfo()
		if opts.CommitSha == "" {
			commitSha = gitInfo.CommitSha
		} else {
			commitSha = opts.CommitSha
		}
		if opts.Branch == "" {
			branch = gitInfo.Branch
		} else {
			branch = opts.Branch
		}
		if opts.RepoURL == "" {
			repoURL = gitInfo.RepoURL
		} else {
			repoURL = opts.RepoURL
		}

		if opts.CommitMessage == "" {
			commitMessage = getCommitMessage()
		} else {
			commitMessage = opts.CommitMessage
		}

		authorName, authorEmail = getAuthorInfo()

		if opts.FilesChanged == nil {
			filesChanged = getChangedFiles()
		} else {
			filesChanged = opts.FilesChanged
		}
	} else {
		commitSha = opts.CommitSha
		branch = opts.Branch
		repoURL = opts.RepoURL
		commitMessage = opts.CommitMessage
		filesChanged = opts.FilesChanged
	}

	if filesChanged == nil {
		filesChanged = []string{}
	}

	// Get diff stats
	additions, deletions := getDiffStats()

	// Send to API
	if t.client.Enabled() {
		t.client.addEvent(map[string]any{
			"type": "git-link-create",
			"body": map[string]any{
				"id":            linkID,
				"traceId":       t.id,
				"observationId": opts.ObservationID,
				"linkType":      string(opts.Type),
				"commitSha":     commitSha,
				"branch":        branch,
				"repoUrl":       repoURL,
				"commitMessage": commitMessage,
				"authorName":    authorName,
				"authorEmail":   authorEmail,
				"filesChanged":  filesChanged,
				"additions":     additions,
				"deletions":     deletions,
				"timestamp":     now.Format(time.RFC3339Nano),
			},
		})
	}

	return &GitLinkInfo{
		ID:            linkID,
		TraceID:       t.id,
		ObservationID: opts.ObservationID,
		Type:          opts.Type,
		CommitSha:     commitSha,
		Branch:        branch,
		RepoURL:       repoURL,
		CommitMessage: commitMessage,
		AuthorName:    authorName,
		AuthorEmail:   authorEmail,
		FilesChanged:  filesChanged,
		CreatedAt:     now,
	}
}

func getCommitMessage() string {
	if out, err := exec.Command("git", "log", "-1", "--format=%s").Output(); err == nil {
		return strings.TrimSpace(string(out))
	}
	return ""
}

func getAuthorInfo() (name, email string) {
	if out, err := exec.Command("git", "log", "-1", "--format=%an").Output(); err == nil {
		name = strings.TrimSpace(string(out))
	}
	if out, err := exec.Command("git", "log", "-1", "--format=%ae").Output(); err == nil {
		email = strings.TrimSpace(string(out))
	}
	return
}

func getChangedFiles() []string {
	files := []string{}
	seen := make(map[string]bool)

	// Staged files
	if out, err := exec.Command("git", "diff", "--cached", "--name-only").Output(); err == nil {
		for _, f := range strings.Split(strings.TrimSpace(string(out)), "\n") {
			if f != "" && !seen[f] {
				files = append(files, f)
				seen[f] = true
			}
		}
	}

	// Unstaged files
	if out, err := exec.Command("git", "diff", "--name-only").Output(); err == nil {
		for _, f := range strings.Split(strings.TrimSpace(string(out)), "\n") {
			if f != "" && !seen[f] {
				files = append(files, f)
				seen[f] = true
			}
		}
	}

	// Untracked files
	if out, err := exec.Command("git", "ls-files", "--others", "--exclude-standard").Output(); err == nil {
		for _, f := range strings.Split(strings.TrimSpace(string(out)), "\n") {
			if f != "" && !seen[f] {
				files = append(files, f)
				seen[f] = true
			}
		}
	}

	return files
}

func getDiffStats() (additions, deletions int) {
	if out, err := exec.Command("git", "diff", "--shortstat").Output(); err == nil {
		output := string(out)
		// Parse: "2 files changed, 10 insertions(+), 5 deletions(-)"
		insertRe := regexp.MustCompile(`(\d+) insertion`)
		deleteRe := regexp.MustCompile(`(\d+) deletion`)

		if match := insertRe.FindStringSubmatch(output); len(match) > 1 {
			additions, _ = strconv.Atoi(match[1])
		}
		if match := deleteRe.FindStringSubmatch(output); len(match) > 1 {
			deletions, _ = strconv.Atoi(match[1])
		}
	}
	return
}
